utilities/blob_db/blob_db.cc

```diff
@@ -17,194 +20,139 @@
-  if (!s.ok()) {
-    return s;
-  }
-  BlockHandle handle;
-  Slice index_entry_slice(index_entry);
-  uint64_t file_number;
-  if (!GetVarint64(&index_entry_slice, &file_number)) {
-    return Status::Corruption();
-  }
-  assert(file_number == 0);
-  s = handle.DecodeFrom(&index_entry_slice);
-  if (!s.ok()) {
-    return s;
+Status BlobDB::Open(const Options& options, const BlobDBOptions& bdb_options,
+                    const std::string& dbname, BlobDB** blob_db) {
+  *blob_db = nullptr;
+  Options myoptions(options);
+  FlushBegin_t fblistener = std::make_shared<BlobDBFlushBeginListener>();
+  ReconcileWal_t rw_filter = std::make_shared<BlobReconcileWalFilter>();
+  EvictAllVersions_t comp_filt_factory =
+      std::make_shared<EvictAllVersionsFilterFactory>();
+
+  myoptions.listeners.emplace_back(fblistener);
+  myoptions.wal_filter = rw_filter.get();
+  myoptions.compaction_filter_factory = comp_filt_factory;
+
+  {
+    MutexLock l(&listener_mutex);
+    all_flush_begins.push_back(fblistener);
+    all_wal_filters.push_back(rw_filter);
   }
-  Footer footer(0, kBlockBasedTableVersionFormat);
-  BlockContents contents;
-  s = ReadBlockContents(file_reader_.get(), footer, options, handle, &contents,
-                        ioptions_);
+
+  DBOptions db_options(myoptions);
+  EnvOptions env_options(db_options);
+
+  // we need to open blob db first so that recovery can happen
+  BlobDBImpl* bdb =
+      new BlobDBImpl(dbname, bdb_options, myoptions, db_options, env_options);
+  fblistener->setImplPtr(bdb);
+  rw_filter->setImplPtr(bdb);
+  comp_filt_factory->setImplPtr(bdb);
+
+  Status s = bdb->openPhase1();
+  if (!s.ok()) return s;
+
+  DB* db = nullptr;
+  s = DB::Open(myoptions, dbname, &db);
+  if (!s.ok()) return s;
+
+  // set the implementation pointer
+  s = bdb->LinkToBaseDB(db);
   if (!s.ok()) {
-    return s;
-  }
-  Block block(std::move(contents), kDisableGlobalSequenceNumber);
-  BlockIter bit;
-  InternalIterator* it = block.NewIterator(nullptr, &bit);
-  it->SeekToFirst();
-  if (!it->status().ok()) {
-    return it->status();
+    delete bdb;
+    bdb = nullptr;
   }
-  *value = it->value().ToString();
+  *blob_db = bdb;
   return s;
 }
+
+BlobDB::BlobDB(DB* db) : StackableDB(db) {}
+
+////////////////////////////////////////////////////////////////////////////////
+//
+//
+// std::function<int(double)> fnCaller =
+//     std::bind(&A::fn, &anInstance, std::placeholders::_1);
+////////////////////////////////////////////////////////////////////////////////
+BlobDBOptions::BlobDBOptions()
+    : blob_dir("blob_dir"),
+      path_relative(true),
+      is_fifo(false),
+      blob_dir_size(1000ULL * 1024ULL * 1024ULL * 1024ULL),
+      ttl_range(3600),
+      min_blob_size(512),
+      bytes_per_sync(0),
+      blob_file_size(256 * 1024 * 1024),
+      num_concurrent_simple_blobs(4),
+      deletion_check_period(2 * 1000),
+      gc_file_pct(20),
+      gc_check_period(60 * 1000),
```

Comment:
RocksDB checks whether a compaction is needed after a flush or a compaction finishes. Can we do the same here? Do we need a time-based GC?